src/Features/CSharp/Portable/DocumentSymbols/CSharpDocumentSymbolsService.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.DocumentSymbols;
+using Microsoft.CodeAnalysis.Host.Mef;
+
+namespace Microsoft.CodeAnalysis.CSharp.DocumentSymbols
+{
+    [ExportLanguageService(typeof(IDocumentSymbolsService), LanguageNames.CSharp), Shared]
+    internal class CSharpDocumentSymbolsService : AbstractDocumentSymbolsService
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CSharpDocumentSymbolsService()
+        {
+        }
+
+        protected override ISymbol? GetSymbol(SemanticModel semanticModel, SyntaxNode node, CancellationToken cancellationToken)
+            => node switch
+            {
+                NamespaceDeclarationSyntax or GlobalStatementSyntax => null,
+                MemberDeclarationSyntax m => semanticModel.GetDeclaredSymbol(m, cancellationToken),
+                LabeledStatementSyntax l => semanticModel.GetDeclaredSymbol(l, cancellationToken),
+                SingleVariableDesignationSyntax s => semanticModel.GetDeclaredSymbol(s, cancellationToken),
+                VariableDeclaratorSyntax v => semanticModel.GetDeclaredSymbol(v, cancellationToken),
+                LocalFunctionStatementSyntax l => semanticModel.GetDeclaredSymbol(l, cancellationToken),
+                TypeParameterSyntax t => semanticModel.GetDeclaredSymbol(t, cancellationToken),
+                _ => null,
+            };
+
+        protected override bool ShouldSkipSyntaxChildren(SyntaxNode node, DocumentSymbolsOptions options)
+        {
+            // If we're not looking for the full hierarchy, we don't care about things nested inside type members
+            if (options == DocumentSymbolsOptions.TypesAndMethodsOnly)
+            {
+                return node is BaseMethodDeclarationSyntax or BasePropertyDeclarationSyntax
+                            or BaseFieldDeclarationSyntax or StatementSyntax or ExpressionSyntax;
```

Comment:
it's a bit odd that it says 'TypesAndMethods', but talks about statements/expressions.  how do those show as hte parent if we're only considering types/methods?

Reply:
If we're only talking about types and methods, then there's no need to delve inside members. You can run into statements, such as at the top level, but if we see one we definitely know that there's nothing of interest inside it. I should also rename that enum to `TypesAndMembers`, it's not method specific.